darwin/Platform.c

```diff
@@ -120,21 +120,65 @@ const MeterClass* const Platform_meterTypes[] = {
    NULL
 };
 
-static double Platform_nanosecondsPerMachTick = 1.0;
+static void Platform_getCPUBrandString(char *cpu_brand_string, size_t cpu_brand_string_size) {
+   if (sysctlbyname("machdep.cpu.brand_string", cpu_brand_string, &cpu_brand_string_size, NULL, 0) == -1) {
+      CRT_fatalError("Could not determine what CPU we are running on");
+   }
+}
 
-static double Platform_schedulerTicksPerNS = -1;
+// Adapted from https://developer.apple.com/documentation/apple-silicon/about-the-rosetta-translation-environment
+static bool Platform_isRunningUnderRosetta2() {
+   int ret = 0;
+   size_t size = sizeof(ret);
+   if (sysctlbyname("sysctl.proc_translated", &ret, &size, NULL, 0) == -1) {
+      if (errno == ENOENT) return 0;
+      CRT_fatalError("Could not determine if this process was running in a translation environment like Rosetta 2.");
+   }
+   return ret;
+}
 
-void Platform_init(void) {
+static double Platform_calculateNanosecondsPerMachTick() {
    // Check if we can determine the timebase used on this system.
    // If the API is unavailable assume we get our timebase in nanoseconds.
-#ifdef HAVE_MACH_TIMEBASE_INFO
-   mach_timebase_info_data_t info;
-   mach_timebase_info(&info);
-   Platform_nanosecondsPerMachTick = (double)info.numer / (double)info.denom;
-#else
-   Platform_nanosecondsPerMachTick = 1.0;
+#ifndef HAVE_MACH_TIMEBASE_INFO
+   return 1.0;
 #endif
 
+   mach_timebase_info_data_t info;
+
+   /* WORKAROUND for `mach_timebase_info` giving incorrect values on M1 under Rosetta 2.
+    *    rdar://FB9546856 https://openradar.appspot.com/radar?id=5055988478509056
+    *
+    *    We don't know exactly what feature/attribute of the M1 chip causes this mistake under Rosetta 2.
+    *    Until we have more Apple ARM chips to compare against, the best we can do is special-case
+    *    the "Apple M1" chip specifically when running under Rosetta 2.
+    */
+
+   size_t cpu_brand_string_size = 1024;
+   char cpu_brand_string[cpu_brand_string_size];
+   Platform_getCPUBrandString(cpu_brand_string, cpu_brand_string_size);
+
+   bool is_running_under_rosetta_2 = Platform_isRunningUnderRosetta2();
+
+   if (is_running_under_rosetta_2 && String_eq(cpu_brand_string, "Apple M1")) {
```

Comment:
Instead of exposing details of Rosetta, translation, and individual CPU brands, I think it'd be cleaner to have a function like Platform_needTimebaseInfoWorkaround() or some such name, and keep all the details about detecting the need for the WAR inside that one function.

Reply:
That's what I initially had, but it's didn't work out great.

We don't know exactly what the nature of the 125/3 factor is, whether it's specific to M1 and will be different with other ARM chips, or generalized to all ARM chips. It's possible that in the future there will be multiple hard-coded WAR factors, and that we'd need to select the appropriate one.

The `Platform_needTimebaseInfoWorkaround` function would check for the various affected hardware configurations, but once you got back a `true`, you'd have to check again to see which particular hardware config you have, so you can apply the right hard-coded factor

Reply:
> WDYT?

Yep - that's pretty much what I was going to suggest.  We could alternatively define a new structure that contains just the workaround information (correction factor) htop needs rather than using the mach structure.  The advantage is it keeps all the low-level platform/hardware/emulation specific details in one place, and not leaking out for the next person hacking here to have to digest, exposing only the minimal information needed by the rest of the code.

Reply:
Actually, I did also try that (I played around with a lot of variants, and forgot about a lot of them hah). It ends up in this weird situation:

```c
mach_timebase_info_t info;

if (Platform_needTimebaseInfoWorkaround(&info)) {
   // Nothing actually goes here, the check also did the setting of `info` :|
} else {
   mach_timebase_info(&info);
}
```

Of course I could simplify it down to:

```c
if (!Platform_needTimebaseInfoWorkaround(&info)) {
  mach_timebase_info(&info);
}
```

But that feels weird, because we're no longer special-casing the workaround, we're special-casing the *lack* of workaround.

Boy if only I could use exceptions, tuples or optionals 😄

Reply:
I'll rework this once new ARM macs come out, and we know whether they're all effected equally, or each need to support different workaround constants.